### PR TITLE
Significantly speed up xpath and class chain queries

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
1lsytv

### CODE REVIEW

**Comment by marekcirkos** [comment]
Looks interesting, can summarize what you changed?

**Comment by mykola-mokhnach** [comment]
Sure @marekcirkos . The main change is, that I've finally found a way to match XCElement to XCUIElementSnapshot without using **_matchesElement** method. 
The secret lies in wdUid property of FBElement, which stays constant for all the snapshots of the same element. So, instead of resolving all the elements and comparing them to each snapshot explicitly one can now simply get uids of matched snapshots and pass them directly to descendants search function using predicate, (see **fb_filterDescendantsWithSnapshots** method of FBUtilities extension), which is times faster.

File: WebDriverAgent.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -34,6 +34,8 @@
 		71A7EAF91E224648001DA4F2 /* FBClassChainQueryParser.h in Headers */ = {isa = PBXBuildFile; fileRef = 71A7EAF71E224648001DA4F2 /* FBClassChainQueryParser.h */; };
 		71A7EAFA1E224648001DA4F2 /* FBClassChainQueryParser.m in Sources */ = {isa = PBXBuildFile; fileRef = 71A7EAF81E224648001DA4F2 /* FBClassChainQueryParser.m */; };
 		71A7EAFC1E229302001DA4F2 /* FBClassChainTests.m in Sources */ = {isa = PBXBuildFile; fileRef = 71A7EAFB1E229302001DA4F2 /* FBClassChainTests.m */; };
+		71B49EC71ED1A58100D51AD6 /* XCUIElement+FBUid.h in Headers */ = {isa = PBXBuildFile; fileRef = 71B49EC51ED1A58100D51AD6 /* XCUIElement+FBUid.h */; };
```
Comment (marekcirkos) [inline WebDriverAgent.xcodeproj/project.pbxproj]: `XCUIElement+FBUID` ?

File: WebDriverAgentLib/Categories/XCUIElement+FBWebDriverAttributes.m [hunk 2]
```diff
@@ -90,6 +91,11 @@ - (NSString *)wdType
   return [FBElementTypeTransformer stringWithElementType:self.elementType];
 }
 
+- (NSUInteger)wdUid
```
Comment (marekcirkos) [inline WebDriverAgentLib/Categories/XCUIElement+FBWebDriverAttributes.m]: wdUID

File: WebDriverAgentTests/IntegrationTests/XCUIElementHelperIntegrationTests.m [hunk 3]
```diff
@@ -53,43 +53,4 @@ - (void)testElementObstruction
...
-  [allElements addObjectsFromArray:windows];
-  
-  NSSet *byTypes = [FBElementUtils uniqueElementTypesWithElements:allElements];
-  NSDictionary *categorizedDescendants = [self.testedApplication fb_categorizeDescendants:byTypes];
-  XCTAssertEqual(2, [categorizedDescendants count]);
-  XCTAssertEqual([categorizedDescendants[@(XCUIElementTypeButton)] count], [buttons count]);
-  XCTAssertEqual([categorizedDescendants[@(XCUIElementTypeWindow)] count], [windows count]);
-}
-
-- (void)testDescendantsFiltering
```
Comment (marekcirkos) [inline WebDriverAgentTests/IntegrationTests/XCUIElementHelperIntegrationTests.m]: Can you add some tests for `fb_filterDescendantsWithSnapshots`?

File: WebDriverAgentLib/Categories/XCUIElement+FBUid.m [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+@end
+
+
+@implementation XCElementSnapshot (FBUid)
+
+- (NSUInteger)fb_uid
+{
+  return [[self.accessibilityElement.payload valueForKey:@"uid.elementID"] intValue];
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBUid.m]: using **objectForKey** now

File: WebDriverAgent.xcodeproj/project.pbxproj (hunk 1 above)
Comment (mykola-mokhnach) [inline WebDriverAgent.xcodeproj/project.pbxproj]: changed

File: WebDriverAgentLib/Categories/XCUIElement+FBWebDriverAttributes.m (hunk 2 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBWebDriverAttributes.m]: changed

File: WebDriverAgentTests/IntegrationTests/XCUIElementHelperIntegrationTests.m (hunk 3 above)
Comment (mykola-mokhnach) [inline WebDriverAgentTests/IntegrationTests/XCUIElementHelperIntegrationTests.m]: added

**Comment by marekcirkos** [comment]
I can't merge it due to https://travis-ci.org/facebook/WebDriverAgent/jobs/237894669#L721. Can you solve it?

**Comment by mykola-mokhnach** [comment]
Fixed

**Pull request closed without merging**